File: app/src/main/java/com/example/tah/ui/task/TaskRecyclerViewAdapter.java
package com.example.tah.ui.task;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.example.tah.R;
import com.example.tah.models.Task;
import com.example.tah.models.TaskViewModel;
import com.example.tah.ui.main.AddAndDetailsActivity;

import java.util.ArrayList;
import java.util.List;


public class TaskRecyclerViewAdapter extends RecyclerView.Adapter<TaskRecyclerViewAdapter.ViewHolder> {

    private final List<Task> tasks;
    private final List<Integer> checkedTasks;
    private final Context context;
    private final TaskViewModel viewModel;

    public TaskRecyclerViewAdapter(List<Task> items, Context context, TaskViewModel viewModel) {
        tasks = items;
        checkedTasks = new ArrayList<>();
        this.context = context;
        this.viewModel = viewModel;
    }

    public void update(List<Task> tasks){
        this.tasks.clear();
        this.tasks.addAll(tasks);
        this.notifyDataSetChanged();
    }

    public List<Integer> getCheckedTasks(){
        return checkedTasks;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_task, parent, false);
        return new ViewHolder(view, viewModel);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.bind(position);
    }

    @Override
    public int getItemCount() {
        return tasks.size();
    }


    class ViewHolder extends RecyclerView.ViewHolder {
        final TextView nameTV;
        final TextView descriptionTV;
        final CheckBox checkBox;
        private final TaskViewModel viewModel;

        Task task;

        public ViewHolder(View view, TaskViewModel viewModel) {
            super(view);
            this.viewModel = viewModel;
            nameTV = (TextView) view.findViewById(R.id.item_name);
            descriptionTV  = (TextView) view.findViewById(R.id.description);
            checkBox = (CheckBox) view.findViewById(R.id.check_box);
        }

        public void bind(int position) {
            task = tasks.get(position);
            nameTV.setText(task.getName());
            checkBox.setChecked(false);

            if(task.getDescription().isEmpty()){
                descriptionTV.setVisibility(View.GONE);
            }else{
                descriptionTV.setVisibility(View.VISIBLE);
                descriptionTV.setText(task.getDescription());
            }

            setOnClickListeners();

            checkBox.setVisibility(viewModel.getCheckBoxVisibility().getValue());
        }

        private void setOnClickListeners() {
            itemView.setOnClickListener(v -> {
                Intent intent = new Intent(context, AddAndDetailsActivity.class);
                intent.putExtra("fragmentId", Task.Companion.getDetailsView());
                intent.putExtra("taskId", task.getId());
                intent.putExtra("name", task.getName());
                intent.putExtra("description", task.getDescription());
                context.startActivity(intent);
            });

            itemView.setOnLongClickListener(v -> {
                int visibility = checkBox.getVisibility() == View.VISIBLE ? View.GONE : View.VISIBLE;
                viewModel.setCheckBoxVisibility(visibility);

                if(visibility == View.GONE){
                    checkedTasks.clear();
                    viewModel.addToCheckedItems(checkedTasks);
                }

                return true;
            });

            checkBox.setOnClickListener(v -> {
                if(checkBox.isChecked()){
                    checkedTasks.add(task.getId());
                }
                else{
                    checkedTasks.remove(task.getId());
                }
                viewModel.addToCheckedItems(checkedTasks);
            });
        }
    }
}

File: app/src/main/java/com/example/tah/ui/task/TaskRecyclerViewAdapter.kt
package com.example.tah.ui.task

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.example.tah.databinding.ItemTaskBinding
import com.example.tah.models.Task
import com.example.tah.ui.main.AddAndDetailsActivity
import com.example.tah.viewModels.TaskViewModel

class TaskRecyclerViewAdapter(
    private val context: Context,
    private val viewModel: TaskViewModel,
    private val checkedTasks: MutableList<Int> = mutableListOf()
) : RecyclerView.Adapter<TaskRecyclerViewAdapter.ViewHolder>() {

    private val differCallback = object : DiffUtil.ItemCallback<Task>() {
        override fun areItemsTheSame(oldItem: Task, newItem: Task): Boolean {
            return oldItem == newItem
        }

        override fun areContentsTheSame(oldItem: Task, newItem: Task): Boolean {
            return oldItem == newItem
        }
    }

    val differ = AsyncListDiffer(this, differCallback)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = ItemTaskBinding.inflate(inflater, parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(differ.currentList[position])
    }

    override fun getItemCount(): Int = differ.currentList.size


    inner class ViewHolder(
        private val binding: ItemTaskBinding
    ) : RecyclerView.ViewHolder(binding.root){

        private lateinit var task: Task

        fun bind(task: Task){
            this.task = task

            with(binding){
                itemName.text = task.name
                checkBox.isChecked = false
                typeImage.setImageResource(task.imageResource)

                if(task.description.isNullOrEmpty()){
                    description.visibility = View.GONE
                }
                else{
                    description.visibility = View.VISIBLE
                    description.text = task.description
                }

                checkBox.visibility = viewModel.getCheckBoxVisibility().value!!
            }
            setOnClickListeners()
        }

        private fun setOnClickListeners() {
            itemView.setOnClickListener {
                val intent = Intent(context, AddAndDetailsActivity::class.java)
                intent.putExtra("fragmentId", Task.getDetailsView())
                intent.putExtra("taskId", task.taskId)
                intent.putExtra("name", task.name)
                intent.putExtra("description", task.description)
                context.startActivity(intent)
            }

            itemView.setOnLongClickListener {
                val visibility: Int = if(binding.checkBox.visibility == View.VISIBLE)
                    View.GONE else View.VISIBLE

                viewModel.setCheckBoxVisibility(visibility)

                if(visibility == View.GONE){
                    checkedTasks.clear()
                    viewModel.addToCheckedItems(emptyList())
                }

                return@setOnLongClickListener true
            }

            binding.checkBox.setOnClickListener {
                if(binding.checkBox.isChecked){
                    checkedTasks.add(task.taskId!!)
                }else{
                    checkedTasks.remove(task.taskId!!)
                }
                viewModel.addToCheckedItems(checkedTasks)
            }
        }
    }

}
